### DISCUSSION TITLE
Welcome to the Copilot Developer Camp Discussions! 🤗

### DISCUSSION THREAD

**Discussion by aycabas** [microsoft/copilot-camp#328] [General]
## Welcome! 👋
We’re using Discussions as a place to connect with our community. We hope that you:
* Ask questions you’re wondering about.
* Share ideas.
* Engage with other community members.
* Welcome others and be open-minded.

Remember that this is a community we build together! 💜

To get started, comment below with an introduction of yourself and tell us about your experience with Copilot Developer Camp 🏕️.

**Comment by BobGerman** [comment]
Hi everyone and welcome to Copilot Developer Camp!
I've had a great time over the last weeks building this with @aycabas and @rabwill . We'd love to know how it goes for you and if we can answer any questions. Thanks!

**Reply by kurt-rolland** [reply]
Hi Bob,

I am interested in the idea of running a Developer Camp. Is there an event platform that you recommend? I am thinking of using Community Days.

**Comment by rabwill** [comment]
Hi @aycabas and @BobGerman my two-favourite developers 👋🏽
Thank you for starting this thread. 

Hey everyone! We're thrilled to have you dive into these labs.  We've poured a ton of effort into getting this up and running, so we really appreciate your excitement and engagement! 😄

**Comment by kamalshree** [comment]
Hi Everyone, 
Thanks @aycabas @BobGerman  and @rabwill  
I'm excited to join this incredible team and am looking forward to contributing. 👍

**Comment by davrous** [comment]
Hi there!

Thanks for this great opportunity! I'm David Rousset working as a Product Manager on the Teams Toolkit, the VS Code extension helping you to create Declarative Agents or Custom Engine Agents. I'm interested in any feedback that could help us improving our product to improve your dev life! ;)

**Comment by cristianoag** [comment]
Hi @aycabas, @rabwill, and @BobGerman. 

I've been trying with the lab. Wow! That is a fantastic resource, however... there are some things that are still not working. Not speaking about customer or internal test tenants that are not in private preview. But even with the Microsoft tenant the labs cannot be completed fully. Here are my feedback points so far:

1. Lab E2 fails to run the app due to missing capabilities on the manifest. These capabilities are defined in the trey-declarative-copilot.json file, which will only become available after Lab E3, Exercise 1. This occurs because, in Lab E2, the manifest.json file already references the declarativeCopilots extension object, but the associated files won’t exist until Lab E3, Exercise 1. I noticed that the latest file set already removed that reference, but my files still had that (I downloaded a few days ago).
2. Incorrect Env File Naming in Lab E3, Exercise 2: The env.local file mentioned in Lab E3, Exercise 2 is incorrectly referred to as env.env.local. It should simply be env.local.
3. Port Occupation Issue: It's common to receive an error indicating that the ports are occupied. Consider adding a note that if this occurs, the func process needs to be terminated to free up the ports. To do this, press CTRL + SHIFT + ESC, find the func process, and end it. This issue occurred multiple times during the lab.
4. Access Permission Prompt for Declarative Copilot: The first time you run the declarative copilot, a prompt appears asking for permission to allow access. This step is not mentioned in the instructions.
5. For some reason, the tunnel that is created during the build process is not the same tunnel that is being used to execute the queries. As you can see down here, the tunnel created during the process is one and the agent is failing trying to access a different one. That thing started to happen since two days ago.

![image](https://github.com/user-attachments/assets/c870702f-a27f-45ae-8a6a-3299c76f710f)
![image](https://github.com/user-attachments/assets/a97fd434-a642-4a44-a5b9-d44fbc663189)

@luishdemetrio FYI!

**Reply by cristianoag** [reply]
We figured out that for some reason the app is not being refreshed automatically on the MSFT tenant and I had to upload the zip manually. After that the devtunnel is correct and I can activate the API. We also noticed that the code is only working OK with the 24100324902 Teams version. I guess a rollout is happening right now that is updating the clients. @luishdemetrio still has an older version and his DA is not working ok.

**Reply by rabwill** [reply]
Yes this is something we have seen as well. glad it's figured out.

**Comment by Viking5274** [comment]
Hi guys! I'm amazed with this camp, but can you help me please with building my own api plugin for declarative agent, for example I've got an fastapi endpoint which return warhammer 40k info via given fraction name, and i want to write a plugin.

Do you have a step-by-step guide? because i don't understand in LAB E2, how i have to modify trey-plugin.json or ai-plugin.json which is created by default, and how actually all request are made?

**Reply by Viking5274** [reply]
Hi @aycabas, @rabwill, and @BobGerman!

**Reply by rabwill** [reply]
Thanks @Viking5274. I have opened an [issue](https://github.com/microsoft/copilot-camp/issues/546) for a content to support how to modify an existing API.

**Comment by swatiarora11** [comment]
I am able to provision the north wind messaging extension typescript sample (downloaded from Copilot for M365 github repo) in the demo tenant. All functionalities work fine with copilot in teams canvas. While provisioning the python implementation of the same sample using visual studio code teams toolkit, I am getting "We could not find a subscription" error. Can someone help me on this issue?
image
![image](https://github.com/user-attachments/assets/b45d4e2e-05ac-4f85-95f1-59a28711342b)

**Reply by swatiarora11** [reply]
@aycabas, @rabwill, and @BobGerman! Requesting your help here

**Reply by rabwill** [reply]
Hi @swatiarora11 ME are not a priorty at the moment for agentic apps in Copilot camp. Can you tell me why you had to specifically go for ME's and not Custom Engine Agents or Declarative agents?

**Comment by nssidhu** [comment]
How can i setup environment where i can experiment ?
I only have outlook account, My company does not faciliate envrionment to practice this.

**Reply by rabwill** [reply]
Unfortunately at the moment you will need atleast a Copilot Licence to do most pathways here like Extend and Make. We are working to make labs work on Copilot chat, this way you can test easily without Copilot license soon. Stay tuned.

**Reply by rhshanbhag** [reply]
Currently I have completed the extend labs on trial subscription which was enabled based on Annual purchase. I am hoping will complete all the labs before the trail subscription ends and will still not be charged.  Is that correct understanding?

**Reply by rabwill** [reply]
Can you tell me more about your trial plan and the annual purchase @rhshanbhag ?

**Reply by rhshanbhag** [reply]
<img width="281" alt="image" src="https://github.com/user-attachments/assets/15fe4e81-747b-4727-ae58-fe62e483ee8c" />
 
 Do let me know if this is right approach. 

Looks like i have already paid 1 month subscription charge for 1 license due to which i was able to do the lab..

**Comment by mahmoudhhassan** [comment]
**Lab E6 - Add Entra ID authentication: Why 2 Apps registrations are needed?**  

Hi Team,

In Lab E6 - Add Entra ID authentication, you defined two app registrations for the API plugin and the API service. I have a few questions about this:

1- Is this absolutely necessary? The documentation here indicates that one app is required for the API service only https://learn.microsoft.com/en-us/microsoft-365-copilot/extensibility/api-plugin-authentication
2- If necessary, could you explain the architecture and the reasoning behind it?
3- If optional, do we have any pros/cons for each approach?

CC @BobGerman @rabwill @aycabas

BR,
Mahmoud

**Reply by rabwill** [reply]
Hi @mahmoudhhassan ,
It is upto you to have one or two apps. The lab E6b,  assumes you already have a registered application for your API and want to integrate it into the agent as a plugin. If you are starting from scratch you can have just one app registration. The reason is also highlighted in the lab as a note.

**Comment by swatiarora11** [comment]
Have some important query, I am getting stuck.
 
I am building declarative copilot plugin. Whenever i launch my custom copilot project from visual studio code IDE, custom copilot caches old development tunnel and returns 404 - not found while invoking REST APIs. 
 
I have been facing this issue from past 4-5 days. I have tried clearing browsing history and cached data, changed users and changed tenants but nothing seems to be working for me. 
 
Can you suggest some hack/fix for it. Require it urgently, as i am not able to test any of the developed functionality.

**Reply by rabwill** [reply]
Have you tried changing the version of the manifest file on every change you make in code?

**Comment by meladhani01** [comment]
Hi Everyone,
Thanks @aycabas @BobGerman and @rabwill
I'm excited to join this incredible team and am looking forward to contributing.

**Comment by ghost** [comment]
Hi everyone. I am Rafsan and I’m a Solution Architect for Copilot Studio. I am excited to learn from you and share my experience in Copilot Studio and extending M356 Copilot.

**Comment by williammabotjaeng** [comment]
Hi Everyone,

I'm William from South Africa. I'm excited to join this camp and contribute to these great products. Cheers!

**Comment by rhshanbhag** [comment]
Hi Team,

Thank you for the Lab Exercise. It's definitely helping us to understand better how application needs to be built and designed.

In Lab E2 when I execute Step 5: Run the application. I get the below exception 

[2025-03-28T04:28:06.462Z] [Error] - Failed to Execute lifecycle deploy due to failed action: {"name":"Ensure database","uses":"script"}. ScriptExecutionError:Execute command failure: node db-setup.js. Env output: {} Any suggestion on how to address this issue?


![EnsureDBScriptIssue](https://github.com/user-attachments/assets/51809493-8dfd-4474-b1d4-b033fb609ec6)

**Reply by rhshanbhag** [reply]
Fixed this issue by updating the Node JS version to 18.16.0.

**Reply by naqvir90** [reply]
This doesnt work for me and im on version 18.16.0. Any other fix? 
Cant go beyond lab e2 because of this error.

**Comment by recife25** [comment]
sorry - just where does one begin??? I attended an event where - we covered the extended portion of building agents - where the completion of labs e2 - e5 were required (I not sure where was that save???) anyway, It was a series of virtual boxes , url , and vscode by launching a virtual box for playground - hence, dont remember - if I completed e2 -t e5 ? , secondly , the portion of the lab dealing with b1 - b2 need to work on. Here is my asking - at first glance - all the labs seem overwhelming. where do I begin ??? 


Video briefing: Building Agents for Microsoft 365 Copilot - should I be here????

**Reply by rabwill** [reply]
Hi Felipe, it is very normal to be overwhelmed with the labs and the process of building agents. Let's break it down step by step to make it more manageable.
Check this flow and see which path way makes sense for you.
![image](https://github.com/user-attachments/assets/e5a7fab6-2512-41bb-9e0c-9d26d91582fa)


If you have a developer tenant where you have a Copilot license I suggest you start here with these two on the `Extend` path. 
- [prerequisite lab](https://microsoft.github.io/copilot-camp/pages/extend-m365-copilot/00-prerequisites/)
- [geo locator game agent lab](https://microsoft.github.io/copilot-camp/pages/extend-m365-copilot/01-declarative-copilot/)

You could also check out the low code `Makers` path with this lab using [Agent builder](https://microsoft.github.io/copilot-camp/pages/make/agent-builder/)

Once comfortable with `Extend` path you could choose path `Build` as it has more configuration to be done by the developer and may be too overwhelming to start with.

**Comment by recife25** [comment]
Hi Rabia,

Gee - thanks for responding. I gather “overwhelm” is the wrong word, it
should be the fork on the road - where to step in first. Let me explain - I
was in a lab online via webinar. It ran for two hours I gather. I managed
to interact with Bob and Angel (Canada). I worked on the extended portion
of the lab - e2 - e7. The second portion of the lab was from b1 to b2. I
have been looking for a way to know where those files are located. I’m
making it more difficult than it should be. I watch all the videos and
follow all the steps. The extended portion was not fully completed, as we
ran out of time. I guess I shall be redoing it all over again. I’m assuming
eventually I will find my way.

Felipe Ruiz

**Comment by rhshanbhag** [comment]
Stuck at Building Custom Engine Agent - Lab B2: Step 3: Debug your app and chat with your data

I did finish step 1 and step 2. On trying to run step 3 the application always get stuck at  Deploy: [2/2] Generating environment variables... Check [Output panel](command:fx-extension.showOutputChannel) for details.

Output Panel Summary:
![image](https://github.com/user-attachments/assets/b55dd515-8595-45e2-a113-9e8718eec434)


not sure what is the issue any inputs??? On cancelling the deployment i can see the below error message in the output panel

![image](https://github.com/user-attachments/assets/b54c611d-7d0d-4a7a-b690-4c904a21fedd)

**Reply by aycabas** [reply]
Hi, sorry to hear that you are struggling with Lab B2. Would you be able to kill the terminals by selecting bin icon on the right side and kill debugging and then try to debug again?

![Screenshot 2025-04-07 at 8 02 41 AM](https://github.com/user-attachments/assets/0e703b57-1481-4208-b9b2-ff8b31b2c6bd)

**Comment by monty0007** [comment]
Hey, so I want to create a custom bot for my SQL Queries. Can anyone here who has done the same or can help me with that? I tried, but the bot was not connecting properly to MSSQL. It was working fine on SQLite.

**Comment by jayshree2020** [comment]
I have completed the low code Makers path with this lab using [Agent builder](https://microsoft.github.io/copilot-camp/pages/make/agent-builder/)

and completed Extend path too. I was curious to know, where can I claim the badge, can't see any MS form to claim. Thanks for the great step by step guidance and detailed instruction in Copilot Developer Camp. Loved it.

**Reply by gretchunkim** [reply]
I completed the entire course back in May, but I haven’t received any badges yet. @rabwill  Could we please know when we will receive our badges? Thanks!

**Comment by SidhantAcharya922** [comment]
Hi, I am trying to send authenticated user email stating me and other user by using '/' in copilot input box, in my declarative agent but its not giving me the appropriate email,even though i have defined openapi json to include emails from query. While I get the Names properly.but not able to get email.
